doc/design/backfill-starvation-prevention.md

```diff
@@ -0,0 +1,179 @@
+# Backfill and Starvation Prevention
+
+## Motivation
+
+### Backfill
+The following diagrams illustrates why we need to add backfill capability to kube batch scheduler. 
+
+We assume that task queue contains 5 tasks, A1, A2, B, C, and D, among which A1 and A2 belong to job A. We further assume that A1 and A2 each requires 1 CPU, and A1 and A2 have to start at the same time. 
+
+Suppose the cluster has only one node. The node has two CPUs. CPU1 is idle and CPU2 is occupied. In the first round, scheduler allocates CPU1 to task A1. There is no CPU available for task A2. Since A1 and A2 have to start at the same time, A1 has to wait, even though it is allocated with enough CPU. Consequently, CPU1 will remain idle in the first round which is a waste of resource. One objective of this feature is to improve resource utilization by allowing other task to run inside allocated resources, CPU1 in this case, via backfill.
+
+![backfill-use-case](../images/backfill-use-case.png)
+
+### Starvation Prevention
+The following diagram illustrates why we need to add starvation prevention capability into kube batch scheduler.
+
+Suppose task A requires 2 CPUs. In the first round, scheduler notices that the node does not have enough CPU to accommodate task A, therefore the scheduler will skip task A, and allocate CPU1 to task B. In the second round, suppose task B occupies CPU1, and CPU2 becomes available.  In this case, we still only have one idle CPU, therefore task A will be skipped again. Scheduler will allocate the node to task C. If this pattern goes on, task A will be starved. Preventing task A from being starved is another issue we need to address in this project.
+
+![starvation-use-case](../images/starvation-use-case.png)
+
+## Objectives
+* Improve Resource Utilization with Backfill
+  * When we schedule jobs by their priorities, we may run into situations where resources are allocated to tasks of a job which is not ready to run. This behavior leads to suboptimal resource utilization.
+  * One goal of this project is to improve resource utilization with backfill - to run tasks in those allocated, but unused, resources.
+  * If preemption is enabled, backfill tasks will be preempted whenever the job (which lends resource) is ready to run.
```

Comment:
Has `backfill task` been running? If yes, could they be preempted the same way as other running tasks? Is there any reason to distinguish the task is a backfill task or not?

Reply:
We distinguish normal tasks and backfill tasks. Normal tasks run inside idle resources. Backfill tasks run inside resources that are allocated to other jobs. In preemption mode, only backfill tasks will be preempted.

Reply:
If I understand it correctly, `kube-batch` could also preempt normal jobs that are decided by `preemptableFns`, and it is useful, e.g. high priority task could preempt low priority tasks.